src/fft_fftw3_f03.f90

```diff
@@ -291,6 +396,148 @@ function decomp_2d_fft_get_sp()
+
+      if (present(igrid)) then
+
+         ! Safety check
+         if (igrid < 1 .or. igrid > n_grid) then
+            call decomp_2d_abort(__FILE__, __LINE__, igrid, "Invalid value for igrid")
+         end if
+         call fft_engines(igrid)%use_it()
+
+      else
+
+         ! Safety check
+         if (n_grid < 1 .or. .not. allocated(fft_engines)) then
+            call decomp_2d_abort(__FILE__, __LINE__, n_grid, &
+                                 "The FFT module was not initialised")
+         end if
+         call fft_engines(1)%use_it()
+
+      end if
+
+   end subroutine decomp_2d_fft_use_grid
+
+   ! Associate the pointers in the module to the ones in the engine
+   subroutine decomp_2d_fft_engine_use_it(engine, opt_force)
+
+      implicit none
+
+      class(decomp_2d_fft_engine), intent(in), target :: engine
+      logical, intent(in), optional :: opt_force ! Skip safety check
+
+      ! Safety check
+      if (.not. engine%initialised) then
+         if (present(opt_force)) then
+            if (.not. opt_force) then
+               call decomp_2d_abort(__FILE__, __LINE__, 0, "FFT engine is not ready")
+            end if
+         else
+            call decomp_2d_abort(__FILE__, __LINE__, 0, "FFT engine is not ready")
+         end if
+      end if
+
+      plan => engine%plan
+      format => engine%format
+      nx_fft => engine%nx_fft
+      ny_fft => engine%ny_fft
+      nz_fft => engine%nz_fft
+      ph => engine%ph
+      sp => engine%sp
+      wk2_c2c => engine%wk2_c2c
+      wk2_r2c => engine%wk2_r2c
+      wk13 => engine%wk13
+
+   end subroutine decomp_2d_fft_engine_use_it
+
+   ! The external code can obtain direct access to internal FFT engines
+   !
+   ! Warning : The external code should not clean the provided engine
+   !           The subroutine decomp_2d_fft_finalize will do it
+   function decomp_2d_fft_get_engine(igrid)
+
+      implicit none
+
+      type(decomp_2d_fft_engine), pointer :: decomp_2d_fft_get_engine
+      integer, intent(in), optional :: igrid
+
+      if (present(igrid)) then
+
+         ! Safety check
+         if (igrid < 1 .or. igrid > n_grid) then
+            call decomp_2d_abort(__FILE__, __LINE__, igrid, "Invalid value for igrid")
+         end if
+         decomp_2d_fft_get_engine => fft_engines(igrid)
+
+      else
+
+         ! Safety check
+         if (n_grid < 1 .or. .not. allocated(fft_engines)) then
+            call decomp_2d_abort(__FILE__, __LINE__, n_grid, &
+                                 "The FFT module was not initialised")
+         end if
+         decomp_2d_fft_get_engine => fft_engines(1)
+
+      end if
+
+   end function decomp_2d_fft_get_engine
+
+   ! The external code can check if the engine is physical_in_x or physical_in_z
+   function decomp_2d_fft_get_format()
+
+      implicit none
+
+      integer :: decomp_2d_fft_get_format
+
+      decomp_2d_fft_get_format = format
+
+   end function decomp_2d_fft_get_format
```

Comment:
I think a lot of this could be in a common file - it doesn't look specific to the FFTW3 backend?